### ISSUE TITLE
Pop-up window gets behind the opened programs

### ISSUE THREAD

**Issue by Danny3** [evilsocket/opensnitch#796]
Hello!

**Describe the bug**
Pop-up window gets behind the opened programs

Include the following information:
 - OpenSnitch version: v1.6.0-rc.4
 - OS: Debian
 - Version: 12 (Bookworm)
 - Window Manager: KDE Plasma 5.26.5 (on Wayland)
 - Kernel version: 6.1.0-1-amd64  SMP PREEMPT_DYNAMIC Debian 6.1.4-1 (2023-01-07) x86_64 GNU/Linux

1. If  Opensnitch is no freshly instlled delete any rules about KDE's Discover or Qbittorrent
2. Open Discover or Qbittorrent, which will atumatically do some connections triggering the Pop-up window of OpenSnitc to appear

**Expected behavior (optional)**
I expected the Po-up window to create a new rule to appear and it appears correctly, but it appears so fast that Discover or Qbittorrent, which appear later are covering it completely when they  appear as the appear more in foreground than it.
Because of this I have to minimize Discover or Qbittorrent, click on the Allow or Deny button and then unminimize them to start doing what I wanted with them.
It's ok that pop-up window appears so fast, I wouldn't want it slower or delayed so I guess it's normal that it's covered by a program that opens up later, but I wish the pop-up windows would have a rule or something to stay above the program that triggered it so I can just click on one of its button and get by to the programs that I opened, without having to do the minimize / unminimize thing.

**Additional context**
I remember I have seen this problem with other programs too.

And there might be even more problematic to decide on Allow or Deny because, for example Kodi and games open up in fullscreen, so there's no minimize button to quickly get that out of the way to see the pop-up window.
In Kodi you have to either quit it or press the "/" to get into its windowed mode, if you know about this shortcut, then minimize it or move it, press Allow | Deny and come back to Kodi and press the "/" to get back into fullscreen.

With games I assume you have to quit it or alt tab it, I'm not sure now as I don't have any game installed to test.

And there's another case with mixed results, for Stellarium:
1. After starting Stellarium, it's loading pop-up appears.
2. Then Open'snitch's pop-up over it, which is great.
3. Then, whe Steallarium is finishes loading, if you haven't already decided to allow it or not, it it opens as full screen covering OpenSnitch's pop-up window and then you can't decide anymore unless you quit it to see the pop-up window again.

So, if you find a way to make the pop-up window show on top of everything or even better on top of the program that triggered it, it would be great for usability.

Another thing that wold be great for usability is that, when a pop-up window appears in the background behind a full screen movie, game, etc, a short optional sound to be played to hear that a pop-up is waiting for you. Maybe some people would like to enable that.
In some case, like when not playing or watching with somebody else, a game or a movie can be paused and the pop-up window be taken care of.
It happened to me to now know that  pop-up appeared to allow something until I have finished and exited a game.
But at the same time I would not want a pop-up in front of a game or movie that I'm playing.

Not sure if this is helpful or not:

https://forum.qt.io/topic/1368/solved-how-to-make-always-top-window
https://amin-ahmadi.com/2016/06/20/how-to-make-your-window-stay-always-on-top/

Thank you!

**Comment by gustavo-iniguez-goya** [comment]
The pop-ups have the sticky flag to be above all other windows. The problem is on Wayland. On Fedora 37 I've realized that this flag is not set.

Could you launch the GUI as follow and try to reproduce it? `$ QT_QPA_PLATFORM=xcb opensnitch-ui`

**Comment by gustavo-iniguez-goya** [comment]
Regarding moving a window on Wayland (spoiler: not supported):

https://gitlab.freedesktop.org/wayland/wayland/-/issues/110
https://lists.freedesktop.org/archives/wayland-devel/2015-September/024410.html

Simple example:
```python
import sys
from PyQt5.QtWidgets import QApplication, QDialog, QLabel, QDesktopWidget
from PyQt5.QtCore import Qt, QSize

app = QApplication(sys.argv)
d = QDialog()
d.setModal(True)
# right click on the title bar, and check if [x] Stay on top is checked on Wayland.
d.setWindowFlags(Qt.WindowStaysOnTopHint)

d.resize(QSize(400, 400))

# move() not supported on Wayland
# pos = QDesktopWidget().availableGeometry()
# d.move(pos.bottomRight())

QLabel('OpenSnitch rlz', d)
d.show()
sys.exit(app.exec_())
```

**Comment by Danny3** [comment]
> The pop-ups have the sticky flag to be above all other windows. The problem is on Wayland. On Fedora 37 I've realized that this flag is not set.
> 
> Could you launch the GUI as follow and try to reproduce it? `$ QT_QPA_PLATFORM=xcb opensnitch-ui`

I'm testing on Debian 12 + unstable repository and latest version of KDE Plasma, Frameworks, on Wayland.

I exited the UI and started it again from the command line as you said and it works great with Discover and Qbittorrent.

In both cases the OpenSnitch's pop-up appears first like before and when Discover and Qbittorrent fully load they remain in background and not cover OpenSnitch pop-up anymore, exactly as I wanted it.

Where it still doesn't work and probably nothing could be done about it is Stellarium where it goes like this:

1. Click on Stelarium in start menu to open it and its loading pop-up appears showing that it's loading its resources
2.  OpenSnitch's pop-up window appears on top of that, as expected.
3. When Stellarium fully loads its pop-up window behind OpenSnitch's one disappears and a fullscreen layout appears covering OpenSnitch's pop-up window.

And since I don't know how to make it windowed like in Kodi, I have to close it to get it out of the way so I can choose an action in OpenSnitch's pop-up.

Now that I reminded myself about Kodi I have tested it there too and strangely even though Kodi open in full screen mode too, similar to Stellarium, there OpenSnitch's pop-up windows stays on top until ou click one of the two buttons, which makes it diappear and then Kodi gets focused as expected.

So Stellarium is the only one that still doesn't play nice.

While both Stellarium and Kodi seem similar how they open in full screen, AFAIK, Stelarium is based on Qt, while Kodi is no, the difference might still be from there.

**Comment by tavoda** [comment]
> And since I don't know how to make it windowed
ALT + TAB?

**Comment by MexHigh** [comment]
Is there still no way to accomplish this on Wayland? Fedora is Wayland only for years and is going to remove X11 from its default installations.

Now I'm fine with running opensnitch-ui via X11 using this flag, but I cannot find a way to do this automatically without setting `QT_QPA_PLATFORM=xcb` globally. I cannot find where the UI is launched at system start - I can only find the `opensnitchd` systemd unit, but not the ui.

EDIT: Found out that you can just edit the autostart file in `~/.config/autostart/opensnitch_ui.desktop`. Replace the line `Exec=opensnitch-ui` with `Exec=env QT_QPA_PLATFORM=xcb opensnitch-ui`. You can even add `--background` to start the UI hidden.

**Comment by gustavo-iniguez-goya** [comment]
Hi @MexHigh ,

As far as I can tell no. Since I've just ported the GUI to PyQt6 I've tested it again, and the issue remains. https://discuss.kde.org/t/kde-plasma-support-for-qt-windowstaysontophint-flag-in-wayland/3106/2 , https://bugreports.qt.io/browse/QTBUG-73456

On v1.7.x there's a new preference option to configure the QT_QPA_PLATFORM options:

![Image](https://github.com/user-attachments/assets/db352793-916f-4c77-96ef-355c1ece0cc9)

**Comment by MexHigh** [comment]
> On v1.7.x there's a new preference option to configure the QT_QPA_PLATFORM options:

Ohh I didn't knew that. Very handy!

Is there a field (or plans for adding one) to start the UI hidden on first start? I think the `--background` flag can be used for this but it is not added automatically.

**Comment by gustavo-iniguez-goya** [comment]
Yeah, `--background` should be it, but you're right, it's only used one passing it from the cli. We can also add it to the Preferences dialog, if that helps :)

**Comment by MexHigh** [comment]
Well, I'm currently doing it by manually modifying the autostart .desktop file. But if the Qt platform plugin is already there, a Background config option would also fit there.

**Comment by stuaxo** [comment]
Does this need to use XDG Activation protocol to work ?

https://blogs.gnome.org/shell-dev/2024/09/20/understanding-gnome-shells-focus-stealing-prevention/

For now, under wayland in gnome-shell, I've used the workaround of 

`gsettings set org.gnome.desktop.wm.preferences focus-new-windows 'smart'`


(Though this probably won't work forever).


With 

`gsettings set org.gnome.desktop.wm.preferences focus-new-windows 'strict'

The windows reliably appear in the background.

**Comment by gustavo-iniguez-goya** [comment]
After this merge https://gitlab.freedesktop.org/wayland/wayland-protocols/-/commit/81632409f6c80fd01b0f3e0284c7984a9d0b7fd0, maybe we'll be able to re-position the pop-ups freely on the screen again.

**Comment by stuaxo** [comment]
> After this merge https://gitlab.freedesktop.org/wayland/wayland-protocols/-/commit/81632409f6c80fd01b0f3e0284c7984a9d0b7fd0, maybe we'll be able to re-position the pop-ups freely on the screen again.

I'm not so sure.   I think it's only meant to be for programs to reposition their own windows.
